### PR TITLE
Update documentation to v2.0.0

### PR DESCRIPTION
**Pull request by pablogs9**
All other PR related to XRCE v2.0.0 should be targetting this one

### CODE REVIEW

File: docs/client.rst [hunk 1]
```diff
@@ -1071,6 +1100,29 @@ Closes a transport previously opened. `PROTOCOL` can be ``udp``, ``tcp`` or ``se
...
+:comm: [TODO]
+:timeout: [TODO]
+
+------
+
+.. code-block:: c
+
+    bool uxr_ping_agent_attempts(const uxrCommunication* comm, const int timeout, const uint8_t attempts);
+
+[TODO]
```
Comment (pablogs9) [inline docs/client.rst:1120]: @jamoralp

File: docs/client.rst [hunk 2]
```diff
@@ -1012,50 +1045,46 @@ Creates an identifier for reference an entity.
 Transport
 `````````
 These functions are platform dependent.
-The values ``PROFILE_XXX_TRANSPORT`` found into ``client.config`` allow enabling some of them.
 The declaration of these functions can be found in ``uxr/client/profile/transport/`` folder.
```
Comment (FranFin) [inline docs/client.rst:1048]:
```suggestion
The declaration of these functions can be found in the ``uxr/client/profile/transport/`` folder.
```

File: docs/client.rst [hunk 3]
```diff
@@ -1012,50 +1045,46 @@ Creates an identifier for reference an entity.
...
 The common init transport functions follow the next nomenclature.
```
Comment (FranFin) [inline docs/client.rst:1049]:
```suggestion
The common init transport functions follow the nomenclature below.
```

File: docs/getting_started.rst [hunk 4]
```diff
@@ -44,8 +44,7 @@ Before create a Session we need to indicate the transport to use (the *Agent* mu
 .. code-block:: C
 
     uxrUDPTransport transport;
-    uxrUDPPlatform udp_platform;
-    if(!uxr_init_udp_transport(&transport, &udp_platform, UXR_IPv4, "127.0.0.1", "2018"))
+    if(!uxr_init_udp_transport(&transport, UXR_IPv4, "127.0.0.1", "2018"))
```
Comment (FranFin) [inline docs/getting_started.rst:47]:
```suggestion
    if (!uxr_init_udp_transport(&transport, UXR_IPv4, "127.0.0.1", "2018"))
```

File: docs/getting_started.rst [hunk 5]
```diff
@@ -296,7 +295,7 @@ Read Data
 ^^^^^^^^^
 Once we have created a valid `DataReader` entity, we can read data from the DDS Global Data Space using the read operation.
 This operation configures how the *Agent* will send the data to the *Client*.
-Current implementation sends one topic to the *Client* for each read data operation of the *Client*.
+Current implementation sends unlimited topics to the *Client*.
```
Comment (FranFin) [inline docs/getting_started.rst:298]:
```suggestion
The current implementation sends unlimited topics to the *Client*.
```

File: docs/installation.rst [hunk 6]
```diff
@@ -4,28 +4,6 @@ Installation
 =========================
 To compile and install the Client and the Agent modules, we use CMake.
```
Comment (FranFin) [inline docs/installation.rst:5]:
```suggestion
To compile and install the Client and the Agent modules, CMake is used.
```

File: docs/installation.rst [hunk 7]
```diff
@@ -77,9 +59,47 @@ This flag will enable the compilation of the examples when the project is compil
...
+Clone the project from GitHub: ::
+
+    $ sudo apt install git openjdk-8-jdk gradle
+    $ git clone https://github.com/eProsima/Micro-XRCE-DDS-Gen.git
+    $ cd Micro-XRCE-DDS-Gen
+    $ git submodules init
+    $ git submodules update
+    $ gradle build -Dbranch=v1.2.5  
+
+You will have the XRCE-DDS available as: ::
```
Comment (FranFin) [inline docs/installation.rst:78]:
```suggestion
You will have the *Micro XRCE-DDS-Gen* tool available as: ::
```

File: docs/shapes_demo.rst [hunk 8]
```diff
@@ -80,7 +80,7 @@ For example, the following command creates the necessary entities for publishing
 To modify the output default stream, you can change it with `stream <id>`.
 
 The maximum available streams corresponds with the ``CONFIG_MAX_OUTPUT_BEST_EFFORT_STREAMS`` and
```
Comment (FranFin) [inline docs/shapes_demo.rst:82]:
```suggestion
The maximum available streams correspond to the ``CONFIG_MAX_OUTPUT_BEST_EFFORT_STREAMS`` and
```

File: docs/shapes_demo.rst [hunk 9]
```diff
@@ -80,7 +80,7 @@ For example, the following command creates the necessary entities for publishing
...
-``CONFIG_MAX_OUTPUT_RELIABLE_STREAMS`` properties configurable in ``client.config`` file.
+``CONFIG_MAX_OUTPUT_RELIABLE_STREAMS`` properties as CMake argument.
```
Comment (FranFin) [inline docs/shapes_demo.rst:83]:
```suggestion
``CONFIG_MAX_OUTPUT_RELIABLE_STREAMS`` properties as CMake arguments.
```

**Review by FranFin — CHANGES_REQUESTED** [review]

File: docs/client.rst [hunk 10]
```diff
@@ -222,6 +222,18 @@ This function logs in a session, enabling any other XRCE communication with the
 
 ------
 
+.. code-block:: c
+
+    void uxr_create_session_retries(uxrSession* session, size_t retries);
+
+Creates a new session with the *Agent* with retries.
```
[the remark this answers is not in the document]
Comment (FranFin) [reply docs/client.rst:229]:
```suggestion
Attempts to establish a new session on the *Agent* :code:`retries` times
```

File: docs/client.rst [hunk 11]
```diff
@@ -1,1078 +1,325 @@
...
+* ``max_samples``: Maximum number of topics that the Agent can send to the Client.
+* ``min_pace_period``: Minimum elapsed time between two topics deliveries, measured in milliseconds,.
+
+For more information, consult the :ref:`read_access_api` of the :ref:`client_api_label`.
+
+Creation Mode
+-------------
 
-:transport: The transport to close.
+The creation of :ref:`entities_label>` on the *Agent* which can act on behalf of the *Clients*
```
Comment (jamoralp) [inline docs/client.rst:271]:
```suggestion
The creation of :ref:`entities_label` on the *Agent* which can act on behalf of the *Clients*
```

File: docs/client.rst [hunk 12]
```diff
@@ -1,1078 +1,325 @@
...
+It comes with an optional ``control`` argument, that allows the *Client* setting the following parameters:
+
+* ``max_bytes_per_second``: Maximum rate at which data messages may be returned, measured in bytes per secpond.
+* ``max_elapsed_time``: Maximum amount of time that can be spent by the Agent in delivering the topic, measured in seconds.
+* ``max_samples``: Maximum number of topics that the Agent can send to the Client.
+* ``min_pace_period``: Minimum elapsed time between two topics deliveries, measured in milliseconds,.
+
+For more information, consult the :ref:`read_access_api` of the :ref:`client_api_label`.
+
+Creation Mode
```
Comment (jamoralp) [inline docs/client.rst:268]: For me, it is very confusing that this section is called "Creation mode" and the one right after is called "Creation Mode table", yet they verse about very different things. We should come up with a clearer way to re-structure this

**Review by jamoralp — CHANGES_REQUESTED** [review]

File: docs/client.rst [hunk 13]
```diff
@@ -1,1078 +1,325 @@
...
+    As detailed in the :ref:`getting_started_label` section, *participants*, *topics*, *datawriters*, *datareaders*, *requesters* and *repliers* work similarly.
+    *Publishers* and *subscribers*, instead, inherit their XML fields from their associated *dataWriters* and *dataReaders*.
+    
+    Creation by XML has the advantage of being configurable direclty within the *Client* application,
+    but comes with the drawback of offering a very limited set of options as regards the QoS with which the DDS entities
+    profiles can be configured. Indeed, only best-effort or reliable communication streams can be set with this creation mode.
+    In many cases, these QoS configurations alone may not be enough. For these cases, *eProsima Micro XRCE-DDS* allows the users
+    to use the creation by references mode.
+
+References
```
Comment (jamoralp) [inline docs/client.rst:310]: I think having an example here would be appropriate, or at least a link to this reference file: https://github.com/eProsima/Micro-XRCE-DDS-Agent/blob/master/agent.refs

File: docs/agent_api.rst [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+    void add_middleware_callback(const Middleware::Kind& middleware_kind, const middleware::CallbackKind& callback_kind, std::function<void (Args ...)>&& callback_function);
+
+This function exposes the same functionality as the one described in :code:`add_middleware_callback`, for :ref:`agent_instance_class_api`.
+
+.. _server_class_api:
+
+eprosima::uxr::Server
+^^^^^^^^^^^^^^^^^^^^^
+
+.. TODO: document start and stop functions.
```
Comment (FranFin) [inline docs/agent_api.rst:319]: @jamoralp TODO

File: docs/client.rst [hunk 15]
```diff
@@ -1,1078 +1,350 @@
...
-The declaration of these functions can be found in ``uxr/client/core/session/stream_id.h`` and ``uxr/client/core/session/object_id.h``.
+.. note::
+    This feature is only available on Linux.
 
-------
+Framing profile
+^^^^^^^^^^^^^^^
 
-.. code-block:: c
+.. TODO: add info/description
```
Comment (FranFin) [inline docs/client.rst:171]: @pablogs9 TODO

File: docs/installation.rst [hunk 16]
```diff
@@ -47,10 +37,24 @@ On Windows first select the Visual Studio version: ::
...
+:code:`/usr/local/lib` to the dynamic loader-linker directories. ::
 
     sudo ldconfig /usr/local/lib/
 
+.. important::
+    The *eProsima Micro XRCE-DDS Agent* executable comes with a rich CLI.
+    Find out all the options offered by this CLI when running the *Agent* in the :ref:`agent_cli` section of the
+    :ref:`micro_xrce_dds_agent_label` page. 
+
+.. TODO: Create a section on snap installation here.
```
Comment (FranFin) [inline docs/installation.rst:54]: @jamoralp TODO

**Review by FranFin — CHANGES_REQUESTED** [review]

File: docs/client.rst [hunk 17]
```diff
@@ -168,7 +168,7 @@ Find out more in the :ref:`dedicated section <discovery_api>` of the API.
 Framing profile
 ^^^^^^^^^^^^^^^
 
-.. TODO: add info/description
+Framing profile enables :ref:`HDLC Framing <_stream_framing_label>` for using :ref:`stream-oriented transports <_intro_transport>` such as Serial transports or custom transports that required framing.
```
Comment (FranFin) [inline docs/client.rst:171]:
```suggestion
The framing profile enables :ref:`HDLC Framing <stream_framing_label>` for using :ref:`stream-oriented transports <intro_transport>` such as Serial transports or Custom transports that require framing.
```

**Review by FranFin — CHANGES_REQUESTED** [review]

File: docs/agent_api.rst (hunk 14 above)
Comment (jamoralp) [reply docs/agent_api.rst]: done

File: docs/installation.rst (hunk 16 above)
Comment (jamoralp) [reply docs/installation.rst]: done

**Comment by jamoralp** [comment]
@FranFin @pablogs9 do you think it would be good to talk somewhere in the documentation about the available Docker images for Micro XRCE-DDS and the Micro XRCE-DDS Agent? Maybe in the installation section?

**Comment by FranFin** [comment]
@jamoralp Yes, think so.

File: docs/client.rst (hunk 15 above)
Comment (FranFin) [reply docs/client.rst]: @pablogs9 This is done, right?

**Comment by pablogs9** [comment]
Are the dockers automated in dockerhub?

File: docs/installation.rst [hunk 18]
```diff
@@ -49,10 +49,32 @@ Now you have the executable *eProsima Micro XRCE-DDS Agent* installed in your sy
...
 
-.. note::
-    If you wish, you can also install the Micro XRCE-DDS Agent as a `Snap package. <https://snapcraft.io/micro-xrce-dds-agent>`_ To do so, simply execute ``sudo snap install micro-xrce-dds-agent`` in your Linux console. You can add the ``--edge`` flag.
+Installation from Snap package
+******************************
+
+    If you wish, you can also install the *eProsima Micro XRCE-DDS Agent* as a `Snap package. <https://snapcraft.io/micro-xrce-dds-agent>`_
+
+    To do so, simply execute ``sudo snap install micro-xrce-dds-agent`` in your Linux console.
+    This will download the Snap package corresponding to the ``stable`` version, that is, *master* branch on GitHub.
```
Comment (FranFin) [inline docs/installation.rst:60]: *master* branch on GitHub -> the *master* branch on GitHub

File: docs/installation.rst [hunk 19]
```diff
@@ -49,10 +49,32 @@ Now you have the executable *eProsima Micro XRCE-DDS Agent* installed in your sy
...
+
+    You can add the ``--edge`` flag to the installation command, just in case you want to download the Snap image corresponding to the *develop* branch on the upstream source code repository.
```
Comment (FranFin) [inline docs/installation.rst:62]: on the upstream source code repository -> from the upstream source code repository

File: docs/installation.rst [hunk 20]
```diff
@@ -49,10 +49,32 @@ Now you have the executable *eProsima Micro XRCE-DDS Agent* installed in your sy
...
+
+.. _install_agent_docker:
+
+Using the provided Docker image
+*******************************
+
+    The *eProsima Micro XRCE-DDS Agent* can also be launched directly from its dedicated `Docker image. <https://hub.docker.com/r/eprosima/micro-xrce-dds-agent>`_
```
Comment (FranFin) [inline docs/installation.rst:69]: Sentence missing a final full stop.

File: docs/installation.rst [hunk 21]
```diff
@@ -49,10 +49,32 @@ Now you have the executable *eProsima Micro XRCE-DDS Agent* installed in your sy
...
+
+    You can pull the image to your local machine by executing ``docker pull eprosima/micro-xrce-dds-agent:<tag> <<args>>``, with *tag* being one of the following options:
+
+    * ``stable``: `Micro XRCE-DDS Agent master <https://github.com/eProsima/Micro-XRCE-DDS-Agent/tree/master>`_ branch.
+    * ``latest``: `Micro XRCE-DDS Agent develop <https://github.com/eProsima/Micro-XRCE-DDS-Agent/tree/develop>`_ branch)
+    * ``vM.m.p``: `Micro XRCE-DDS Agent tagged <https://github.com/eProsima/Micro-XRCE-DDS-Agent/tags>`_ versions, with the *Major*, *minor*, *patch* structure.
+
+    The accepted arguments for ``<<args>>`` are exactly the same ones which are listed in the :ref:`agent_cli` section.
```
Comment (FranFin) [inline docs/installation.rst:77]: same ones -> same

File: docs/installation.rst [hunk 22]
```diff
@@ -93,6 +115,23 @@ For building your Client app in your host machine, you need to build against the
 
     gcc <your_main.c> -lmicrocdr -lmicroxrcedds_client
 
+.. _install_client_docker:
+
+Using the provided Docker image
+*******************************
+
+    The *eProsima Micro XRCE-DDS Client* comes with a `Docker image <https://hub.docker.com/r/eprosima/micro-xrce-dds-agent>`_ where the library is installed, as well as the provided examples, so users can easily execute them.
```
Comment (FranFin) [inline docs/installation.rst:123]: where the library is installed together with the examples, so they can easily be executed by the users.

File: docs/installation.rst [hunk 23]
```diff
@@ -93,6 +115,23 @@ For building your Client app in your host machine, you need to build against the
...
+
+
+    You can pull the image to your local machine by executing ``docker pull eprosima/micro-xrce-dds-client:<tag> <<args>>``, with *tag* being one of the following options:
+
+    * ``stable``: `Micro XRCE-DDS Client master <https://github.com/eProsima/Micro-XRCE-DDS-Client/tree/master>`_ branch.
+    * ``latest``: `Micro XRCE-DDS Client develop <https://github.com/eProsima/Micro-XRCE-DDS-Client/tree/develop>`_ branch)
+    * ``vM.m.p``: `Micro XRCE-DDS Client tagged <https://github.com/eProsima/Micro-XRCE-DDS-Client/tags>`_ versions, with the *Major*, *minor*, *patch* structure.
+
+    The accepted arguments for ``<<args>>`` are the examples' executable names, along with the arguments required for each example to work.
```
Comment (FranFin) [inline docs/installation.rst:132]: along with -> followed by

File: docs/installation.rst [hunk 24]
```diff
@@ -139,4 +178,25 @@ Usually is useful to install examples along with the XRCE-DDS suite, for doing s
...
-    to the :code:`cmake ..` command-line options. This flag will enable the compilation of the examples.
\ No newline at end of file
+    to the :code:`cmake ..` command-line options. This flag will enable the compilation of the examples.
+
+.. _install_agent_client_docker:
+
+Using the provided Docker image
+*******************************
+
+    *eProsima Micro XRCE-DDS* is also available as a whole package in a `Docker image. <https://hub.docker.com/r/eprosima/micro-xrce-dds-agent>`_
```
Comment (FranFin) [inline docs/installation.rst:188]: Missing full stop at the end.

File: docs/installation.rst [hunk 25]
```diff
@@ -139,4 +178,25 @@ Usually is useful to install examples along with the XRCE-DDS suite, for doing s
...
+
+    Within this Docker image, the *Micro XRCE-DDS Agent* standalone application and library are installed, as well as the *Micro XRCE-DDS Client* library and built-in examples.
```
Comment (FranFin) [inline docs/installation.rst:190]: Within -> With

**Review by FranFin — CHANGES_REQUESTED** [review]

File: docs/installation.rst (hunk 20 above)
Comment (jamoralp) [reply docs/installation.rst:69]: The full stop is included in the html reference :)

File: docs/installation.rst (hunk 24 above)
Comment (jamoralp) [reply docs/installation.rst:188]: Same as before, full stop included in the html reference

File: docs/installation.rst (hunk 25 above)
Comment (jamoralp) [reply docs/installation.rst:190]: well, actually the installation is done inside the Docker. Wouldn't it make more sense to keep it as within?

File: docs/index.rst [hunk 26]
```diff
@@ -2,128 +2,122 @@
...
+This example is useful to understand how the DDS-XRCE protocol interfaces with the DDS World.
 
-To create your client, you can follow the instructions of :ref:`getting_started_label` page.
-It is a tutorial that describes briefly how *eProsima Micro XRCE-DDS* API is and how it works.
+To learn how to handle all the ingredients needed to create a *Client*, carefully read the :ref:`getting_started_label` page.
+This page describes how to use the *eProsima Micro XRCE-DDS* API in order to set up and run a *Client* application.
 
-To know more about *eProsima Micro XRCE-DDS*, you can read the corresponding parts to the :ref:`micro_xrce_dds_client_label` or the :ref:`micro_xrce_dds_agent_label`.
-If you are interested in how XRCE works, read :ref:`entities_label` and :ref:`operations_label` pages.
+A genric introduction to the library can be found in the :ref:`user` page.
```
Comment (jamoralp) [inline docs/index.rst:82]:
```suggestion
A generic introduction to the library can be found in the :ref:`user` page.
```

File: docs/optimization.rst [hunk 27]
```diff
@@ -3,42 +3,55 @@
...
-   but for the majority of purposes, only one stream of either best effort or reliable can be used.
 
-2. The history of a reliable stream is used for recovering from lost messages and speeding up the communication.
-   For output streams, more history value will allow writing and sending more messages without waiting for confirmation.
-   If a history of an output stream is full (no messages were still confirmed by the *Agent*), 
-   no more messages can be stored in the stream.
-   For input streams, the history is used for recovering faster when the messages are lost using less extra band-width.
-   If your connection is highly reliable and to save memory is a priority, a reduced history can be used.
+Streams number
+   It's possile to define a maximum of *127* best-effort streams and of *128* reliable streams.
```
Comment (jamoralp) [inline docs/optimization.rst:35]:
```suggestion
   It's possible to define a maximum of *127* best-effort streams and of *128* reliable streams.
```

**Review by jamoralp — APPROVED** [review]

**Review by FranFin — APPROVED** [review]

**Pull request merged**